### PR TITLE
Add JMH microbenchmarks

### PR DESCRIPTION
**Pull request by szpak**
### General description

As discussed with @bsideup at GeeCON, I benchmarked the inpact of Bloud Hound integration to check how (un)safe it could be to enable it in production. I started with the following cases:
 - non blocking Mono call (simple block() on the `single()` scheduler) 
 - non blocking Flux call (100K elements, simple blockLast() on the `single()` scheduler)
 - blocking Mono call (simple block() on the `single()` scheduler)
 - blocking Flux call (100K elements, simple blockLast() on the `single()` scheduler)

I tried to provide a baseline to a variant without the integration enabled (installed). As I haven't found a sensible reason to test blocking Flux which throws exception and aborts execution, so I applied a custom `blockingMethodCallback()` to simulate a warning printed on every blocking method detection (e.g. during the execution in production).

In my tests, I took the worst scenario where there is no custom logic in Flux elements processing. In many cases the relative Block Hound impact will be lower.

All benchmarks are performed in a single thread environment. It could be potentially extended to execute multiple streams processing in multiple thread, but then more external factors can make the results less reliable.

### Some Reactor-related question

Talking about reliability and reproducibility ocasionally I've been getting some variation in results - even for Reactor itself (without Block Hound). For example, for the naive baseline benchmark:

```
    @Benchmark
    public Integer baselineNonBlockingFluxSubscribedOnSingle() {
        return numbers.hide().subscribeOn(single()).blockLast(Duration.ofSeconds(1));
    }
```

with numbers being non-blocking Flux`numbers = Flux.range(0, 100_000);` I receive result with quite large standard deviation (`~20%` of the average). It's all with Block Hound disabled. I chose `single()` to play with just one thread. 

```
Result "reactor.blockhound.BlockHoundNonBlockingBenchmark.baselineNonBlockingFluxSubscribedOnSingle":
  1140.786 ±(99.9%) 233.514 us/op [Average]
  (min, avg, max) = (981.872, 1140.786, 1448.813), stdev = 218.429
  CI (99.9%): [907.272, 1374.299] (assumes normal distribution)

Benchmark                                                                 Mode  Cnt     Score     Error  Units
BlockHoundNonBlockingBenchmark.baselineNonBlockingFluxSubscribedOnSingle  avgt   15  1140.786 ± 233.514  us/op
```

It flattens nicely with multiple JVM forks, but anyway having:

```
# Fork: 1 of 3
Iteration   1: 1927.348 us/op
Iteration   2: 1906.449 us/op
Iteration   3: 1922.533 us/op
...

# Fork: 2 of 3
Iteration   1: 1310.879 us/op
Iteration   2: 1313.213 us/op
Iteration   3: 1329.674 us/op
```

for the same operation puzzles me a little bit. It can be an impact of my environment, but **maybe you have some idea what in Reactor itself could make the passes so distinct? Maybe there is something I could tune in scheduling (or something else) to make the result more similar?**

### Summary

To conclude, I've got some raw results, but before posting it here and drawing some conclusions it would be good to test more cases where you suspect Bloud Hound impact could be possibly high (and also repeat my tests with higher number of iterations/forks). **@bsideup what would you like to test?**

*Btw, the tests can be execute with `gw jmh` or (with some minimal configuration tuning) with the JMH plugin for Idea (less accurate, but more convinient when working of test cases).*

### CODE REVIEW

**Comment by bsideup** [comment]
Hi @szpak,

Thanks a lot for finding time to work on it! 👍

Since BlockHound works with any JVM code, not just Reactor, I believe Reactor is not necessary here. We should measure the overhead of calling a blocking method in the following scenarios:

1. blocking-friendly thread
1. non-blocking thread, no "markers" in the stacktrace (aka `allowBlockingCallsInside`)
1. non-blocking thread, but `allowBlockingCallsInside` in the stack
1. non-blocking thread, but `disallowBlockingCallsInside` in the stack

Also, make sure that you override the callback (it throws an exception by default, we don't need it in the benchmarks)

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,13 @@
 services:
   - docker
 
+script:
```
Comment (bsideup) [inline .travis.yml]: let's not run the benchmarks during the CI

File: example/build.gradle [hunk 2]
```diff
@@ -47,4 +48,25 @@ dependencies {
     reactor_3_3_x 'io.projectreactor:reactor-core:3.3.0.BUILD-SNAPSHOT', {
         changing = true
     }
+
+    jmh project(path: ":agent", configuration: 'shadow')
+    //Uncomment for benchmark execution by Idea plugin
+//    jmh "org.openjdk.jmh:jmh-generator-annprocess:1.21"
```
Comment (bsideup) [inline example/build.gradle]: I guess we can keep it uncommented, it won't hurt, but less manual actions to run it from IDEA after a fresh checkout

File: example/build.gradle [hunk 3]
```diff
@@ -47,4 +48,25 @@ dependencies {
...
+}
+
+jmh {
+    if (project.hasProperty("jmh.simpleTravisMode")) { //temporary for JHM sanity testing in Travis
```
Comment (bsideup) [inline example/build.gradle]: leftover?

File: example/build.gradle [hunk 4]
```diff
@@ -47,4 +48,25 @@ dependencies {
...
+        include = 'reactor\\.blockhound\\.BlockHoundDummyBenchmark.*'
+        warmupIterations = 1
+        iterations = 3
+        fork = 1
+    } else {
+        include = 'reactor\\.blockhound\\.BlockHound.*Benchmark.*'
+        warmupIterations = 3
```
Comment (bsideup) [inline example/build.gradle]: I would prefer if there are configured as an annotation on the concrete test, because different benchmarks may use different parameters

File: example/build.gradle [hunk 5]
```diff
@@ -47,4 +48,25 @@ dependencies {
...
+        iterations = 5
+        fork = 3
+    }
+    resultsFile = project.file("${project.buildDir}/reports/jmh/results${System.currentTimeMillis()}.txt")
```
Comment (bsideup) [inline example/build.gradle]: it will make it harder to work with (a lot of files created), let's remove the timestamp from the file's name

File: example/src/jmh/java/reactor/blockhound/BlockHoundBlockingBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
+package reactor.blockhound;
```
Comment (bsideup) [inline example/src/jmh/java/reactor/blockhound/BlockHoundBlockingBenchmark.java]: License header is missing

**Comment by szpak** [comment]
Thanks @bsideup for your comments. I will take a position on them after Devoxx Poland (which starts the following week).

File: .travis.yml (hunk 1 above)
Comment (szpak) [reply .travis.yml]: I had some issues with C compilation locally and I wanted to test it on Travis. However, I plan to remove it before the merge.

File: example/build.gradle (hunk 2 above)
Comment (szpak) [reply example/build.gradle]: Unfortunately not. The plugin adds it itself and there is an error about duplicated generated classes. There is a possibility of changing `duplicateClassesStrategy`, but it can have side effects.
It would be best to detect import from Idea and only then add that dependency, but I haven't found any good way to do that.

File: example/build.gradle (hunk 3 above)
Comment (szpak) [reply example/build.gradle]: No, it's on purpose in the first stage - due to aforementioned local issues with C compilation.

File: example/build.gradle (hunk 4 above)
Comment (szpak) [reply example/build.gradle]: Currently it's not a case, but I removed that from `build.gradle`.

File: example/build.gradle (hunk 5 above)
Comment (szpak) [reply example/build.gradle]: It depends what you plan to achieve. But in the end - the stable state - one file is ok.

**Comment by szpak** [comment]
I've reworked the benchmarks to cover mentioned cases. I removed all reference to Reactor and changed (potentially) blocking `yield()` to a custom non-blocking method marked as blocking (due to #39 I have to use existing method from JDK).

Because of #38 it is problematic to have only non-blocking calls with BlockHound installed from a thread that should be used for non-blocking operations. Without `ThreadPoolExecutor` reusing one thread across all measured calls is problematic.
For the same reason a baseline for non-blocking calls is not fully reliable as there is a hidden blocking call. Nevertheless, it shouldn't make real impact.

I made benchmark execution with 5 iterations (+ 3 warmup iterations) in 3 forks. The full results are [there](https://gist.github.com/szpak/14c28b91b0ce66918e1b8651339d8722). The summary is here:
```
Benchmark                                                          Mode  Cnt  Score   Error  Units
BHBlockingAllowedThreadBenchmark.baselineBlockingCall              avgt   15  5.124 ± 0.039  us/op
BHBlockingAllowedThreadBenchmark.measureBlockingCall               avgt   15  5.222 ± 0.078  us/op
BHBlockingAllowedThreadBenchmark.baselineNonBlockingCall           avgt   15  5.070 ± 0.045  us/op
BHBlockingAllowedThreadBenchmark.measureNonBlockingCall            avgt   15  5.142 ± 0.076  us/op
BHBlockingNotAllowedThreadBenchmark.baselineBlockingCall           avgt   15  5.102 ± 0.067  us/op
BHBlockingNotAllowedThreadBenchmark.measureAllowedBlockingCall     avgt   15  5.853 ± 0.184  us/op
BHBlockingNotAllowedThreadBenchmark.measureDisallowedBlockingCall  avgt   15  6.334 ± 0.340  us/op
```
(I rearranged order in the output and shortened `BlockHound` to `BH` for readibility)

As expected, the impact on "blocking-friendly" threads is negligible (4 four lines). Difference between `baselineBlockingCall` and `baselineNonBlockingCall` in `BHBlockingAllowedThreadBenchmark` itself is caused by slightly simpler call.
In "non-blocking" threads the impact it a little bit bigger - more call related operations are performed. The disallowed blocking call has bigger impact which can be caused by extra handlers involved. Not performed due to #38 check on non blocking calls would probably have slightly smaller impact than "allowed blocking calls" as a short path should be used.
I could run simulate that case by commenting out `Unsafe.park()`, but it would be problematic to reproduce in the future and maybe you know some better way (without writing my own thread executor :) ).

We could also perform some test on more complex call to check how that 10-20% overhead (in a negative case) behaves in a function of number of accompanying non blocking calls.

Nevertheless, please write first how you see those use cases and received numbers.

Btw, once merging, I propose to merge it without squashing to keep a reference to the first version with code for Reactor, just in a case we would like to come back to that idea in the future.

**Comment by bsideup** [comment]
Hi @szpak,

Thanks a lot for your work!I pushed a change to your branch that simplifies the benchmarks and focuses on the actual overhead, I hope you don't mind :) I will appreciate your review in case I accidentally broke something.

Here are the results I got:
```
Benchmark                             Mode  Cnt  Score   Error  Units
baselineNonBlockingCall               avgt   15  0,066 ± 0,001  us/op
measureNonBlockingCall                avgt   15  0,066 ± 0,001  us/op

baselineBlockingCallInBlockingThread  avgt   15  0,489 ± 0,006  us/op
measureBlockingCallInBlockingThread   avgt   15  0,807 ± 0,011  us/op

baselineAllowedBlockingCall           avgt   15  0,495 ± 0,015  us/op
measureAllowedBlockingCall            avgt   15  2,836 ± 0,016  us/op
```

While the percentages may look scary, I would like to focus on the differences instead :)

We have 3 scenarios here:
# 1. non-blocking call
when the executed method is not marked as blocking.

As we can see from the numbers, there is zero overhead from BlockHound on non-blocking methods. Expected, but good to verify.

# 2. blocking call but in non-blocking thread
a regular blocking call inside a thread that is **not** marked as non-blocking.

This one is not a problem yet has to be intercepted, hence the overhead. The benchmark reports an overhead of ~0.3us per blocking method invocation. Given that the call is blocking, most probably it will block for a duration measured by milliseconds, which means that the *total* overhead is low.

# 3. blocking call inside a non-blocking thread but marked as "allowed"

The heaviest one. A blocking call is detected but whitelisted (e.g. some logging). The difference is big (~5.8 times), but again, the overhead is ~2.3us per blocking call.
The idea of having the whitelisting is to be able to ignore some blocking calls that are "accepted", but it also means that such blocking calls should happen rarely and not affect the overall performance. 2us for such calls is IMO an acceptable overhead.

---

The bottom line is: 
1. BlockHound **does add the overhead** (mostly due to the fact that it is a native agent and comes with the overhead of JNI), but only to the blocking methods.
2. Such methods are usually measured in milliseconds (IO, sleeping, awaiting, etc), and BlockHound only adds a few **micro**seconds (1 millisecond = 1000 microseconds) to it.
3. Adding BlockHound should not dramatically reduce the performance, unless the performance is measured at microseconds level. Since the primary goal of BlockHound is to detect the blocking calls during the development & testing, this is more than acceptable. The observed overhead is even acceptable for staging environments.
4. Production usage is possible, but should be considered carefully in case of the performance critical services. Note that this is **not** a regular use case of BlockHound.


P.S. after running the benchmarks with async-profiler, I identified a few possible optimizations in the native agent we may consider (e.g. fail fast if the native method is never marked as "allowed in some cases"). This will require some API changes and should be considered separately.

File: benchmarks/src/jmh/java/reactor/blockhound/BlockHoundBenchmark.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Benchmark
+    public void baselineNonBlockingCall() throws Exception {
+        Thread thread = new Thread(runMultipleTimes(BlockHoundBenchmark::nonBlockingCall));
+        thread.start();
+        thread.join(5_000);
+    }
+
+    @Benchmark
+    public void measureNonBlockingCall(BlockHoundInstalledState state) throws Exception {
```
Comment (szpak) [inline benchmarks/src/jmh/java/reactor/blockhound/BlockHoundBenchmark.java]: According to the JMH documentation calls inside the benchmark that do not return a value are easier to inline the optimalizations in JVM. Therefore, my tests had a return value or were using a `BlackHole` to prevent that. However, I haven't analyzed the calls using threads or sleep(0)/incrementAndGet() with a profiler, so maybe you were able to ensure that it doesn't happen. Similar situation with optimizations is with running it in a loop.

In general, it's nice that you were able to simplify the benchmarks itself.

Comment (bsideup) [reply benchmarks/src/jmh/java/reactor/blockhound/BlockHoundBenchmark.java]: the benchmark starts a Thread now, hence the inlining cannot affect the results I guess

**Comment by bsideup** [comment]
@szpak I merged the PR 🎉 Thanks a lot for the initial implementation and the feedback 👍

While evaluating the benchmarks, I already discovered a few easy ways how to reduce the overhead and will provide follow-up PRs with the optimizations 💯
